packages/openapi-code-generator/src/typescript/common/schema-builders/zod-v4-schema-builder.ts

```diff
@@ -0,0 +1,388 @@
+import type {TypeBuilder} from "../type-builder"
+import {quotedStringLiteral} from "../type-utils"
+import type {ExportDefinition} from "../typescript-common"
+import {
+  AbstractSchemaBuilder,
+  type SchemaBuilderConfig,
+} from "./abstract-schema-builder"
+
+const zod = "z"
+
+export const staticSchemas = {
+  PermissiveBoolean: `${zod}.preprocess((value) => {
+          if(typeof value === "string" && (value === "true" || value === "false")) {
+            return value === "true"
+          } else if(typeof value === "number" && (value === 1 || value === 0)) {
+            return value === 1
+          }
+          return value
+        }, ${zod}.boolean())`,
+  PermissiveLiteralTrue: `${zod}.preprocess((value) => {
+          return PermissiveBoolean.parse(value)
+        }, ${zod}.literal(true))`,
+  PermissiveLiteralFalse: `${zod}.preprocess((value) => {
+          return PermissiveBoolean.parse(value)
+        }, ${zod}.literal(false))`,
+}
+
+export type StaticSchemas = typeof staticSchemas
+
+export class ZodV4Builder extends AbstractSchemaBuilder<
+  ZodV4Builder,
+  StaticSchemas
+> {
+  readonly type = "zod-v4"
+
+  static async fromInput(
+    filename: string,
+    input: Input,
+    schemaBuilderConfig: SchemaBuilderConfig,
+    typeBuilder: TypeBuilder,
+  ): Promise<ZodV4Builder> {
+    return new ZodV4Builder(
+      filename,
+      input,
+      schemaBuilderConfig,
+      typeBuilder,
+      staticSchemas,
+    )
+  }
+
+  override withImports(imports: ImportBuilder): ZodV4Builder {
+    return new ZodV4Builder(
+      this.filename,
+      this.input,
+      this.config,
+      this.typeBuilder,
+      staticSchemas,
+      {},
+      new Set(),
+      imports,
+      this,
+    )
+  }
+
+  protected importHelpers(imports: ImportBuilder) {
+    imports.from("zod/v4").add(zod)
+  }
+
+  public parse(schema: string, value: string): string {
+    return `${schema}.parse(${value})`
+  }
+
+  public schemaTypeForType(type: string): string {
+    return type ? `${zod}.ZodType<${type}>` : ""
+  }
+
+  protected schemaFromRef(reference: Reference): ExportDefinition {
+    const name = getSchemaNameFromRef(reference)
+    const schemaObject = this.input.schema(reference)
+
+    const value = this.fromModel(schemaObject, true)
+
+    let type = ""
+
+    // todo: bit hacky, but it will work for now.
+    if (value.includes("z.lazy(")) {
+      type = this.typeBuilder.schemaObjectToType(reference)
+    }
+
+    return {
+      name,
+      type: this.schemaTypeForType(type),
+      value,
+      kind: "const",
+    }
+  }
+
+  protected lazy(schema: string): string {
+    return [zod, `lazy(() => ${schema})`].join(".")
+  }
+
+  protected merge(schemas: string[]): string {
+    const definedSchemas = schemas.filter(isDefined)
+
+    if (definedSchemas.length === 1 && definedSchemas[0]) {
+      return definedSchemas[0]
+    }
+
+    // todo: .merge is deprecated in v4, plan migration to .extend
```

Comment:
I suspect this is going to be a pain to implement, so punting on it for this initial PR and continuing to use the deprecated method :grimacing: